src/librbd/mirror/GetInfoRequest.cc

```diff
@@ -137,44 +128,121 @@ void GetInfoRequest<I>::handle_get_journal_tag_owner(int r) {
 }
 
 template <typename I>
-void GetInfoRequest<I>::get_snapshot_promotion_state() {
```

Comment:
Initially I was thinking about this too, but it looked the state machine would be too confusing to read though I decided to have  it not optimal but much simpler.

Ok, I will add the optimization.

Reply:
Can definitely drop the `refresh_image` step so that this issue gets fixed [1], so then it could just copy the snapshots at construction time and set a boolean saying it's valid to know if you should skip the look-up states.

[1] https://tracker.ceph.com/issues/43589